src/test/unit/io/deserializer_test.cpp

```diff
@@ -714,6 +714,49 @@ TEST(deserializer_vector, sum_to_zero_jacobian) {
   EXPECT_FLOAT_EQ(lp_ref, lp);
 }
 
+TEST(deserializer_matrix, sum_to_zero_constrain) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(3.0);
+  theta.push_back(-1.0);
+  theta.push_back(-2.0);
+  theta.push_back(0.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  double lp = 0.0;
+  Eigen::MatrixXd reference
+      = stan::math::sum_to_zero_constrain(stan::math::to_matrix(theta, 2, 2));
+  Eigen::MatrixXd phi(
+      deserializer.read_constrain_sum_to_zero<Eigen::MatrixXd, false>(lp, 3,
+                                                                      3));
+  for (int i = 0; i < phi.rows(); ++i) {
```

Comment:
[optional] If `reference` and `phi` are the same size and same row-major vs. column major, it's easier to do this as

```cpp
for (int n = 0; n < phi.size(); ++n) {
  EXPECT_FLOAT_EQ(reference(n), phi(n));
}
```